### ISSUE TITLE
Thoughts on scaling kine

### ISSUE THREAD

**Issue by tamalsaha** [rancher/kine#36]
Thanks for building this @ibuildthecloud ! 

- After reading up on the etcd data model, I wonder if it will be possible to run kine in a HA setup. Have you tried that? Any thoughts?

- Do you think it will be possible to store different data types in different tables so that it can scale well. Also, I am only curious about Postgres. Have you tried using jsonb data type so one can write reach queries against the table directly?

**Comment by liorokman** [comment]
>Do you think it will be possible to store different data types in different tables so that it can scale well. Also, I am only curious about Postgres. Have you tried using jsonb data type so one can write reach queries against the table directly?

If you're only interested in PostgreSQL, you can store different data types in different tables without touching the Kine code. You just need to leverate PostgreSQL's [partitioned tables](https://www.postgresql.org/docs/13/ddl-partitioning.html). You can pre-create the `kine` table as a `range partitioned` table, and create separate partitions for each object type - as encoded in the `name` column. For example, a ranged partition from `/registry/namespaces/` to `/registry/namespaces0` will contain all the namespace objects.

Using jsonb would not work - Kubernetes actually uses a binary protobuf encoding for storing core objects. Only CRD instances are stored as JSON. Not to mention that when the API server is [configured to encrypt a resource](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/) (e.g. a Secret), the object is encrypted, the data stored in the database is encrypted, and therefore would be unreadable, even if it were in JSON.

**Comment by lucasfcnunes** [comment]
Is there a way to decode this binary protobuf encoding to get it all in json?

**Comment by brandond** [comment]
Yes, but that's a question for the Kubernetes project, not Kine.

**Comment by lucasfcnunes** [comment]
I was thinking about using kine to apply resources using SQL. The protobuf encoding is gonna make it difficult to replace values.

**Comment by brandond** [comment]
... what? No, don't go modifying the datastore directly without going through the apiserver.

**Comment by lucasfcnunes** [comment]
Isn't kine used as a disaster recovery solution?
If I restart the cluster isn't it going to use the data store by kine?
My thought process was towards using it to

1. Apply some resources in the cluster
1. Shutdown the cluster down
1. Change things like ingress hosts, configmaps and secrets
1. Start the cluster

This way I could easily replicate, tweak and restore cluster state in isolated environments.

I was even thinking about using it to forbid some patterns to be inserted into the database.
